misc/userspace-task-models.md

```diff
@@ -0,0 +1,240 @@
+
+The most basic task model under these assumptions is a synchronous task model
+where all tasks are scheduled via `scheduler.postTask()` and end when the
+associated callback terminates:
+
+![Simplified Synchronous Task Model](../images/synchronous-tasks-simplified.png)
+
+This task model works well if the tasks are reasonably short, but it can lead
+to unresponsive pages if the tasks are long. The two common approaches to
+mitigate this are:
+
+ 1. Break up the tasks into smaller pieces, scheduling all the pieces up
+    front[<sup>4</sup>](#notes)
+ 1. Yield to the event loop after some time, scheduling a continuation to
+    resume
+
+### Yieldy Asynchronous Tasks
+
+Yieldy async tasks are spread over multiple browser tasks, either by scheduling
+a *continuation task* with `yield()` or `wait()` or by using an async API like
+`fetch()`.
+
+![Simplified Yieldy Asynchronous Task Model](../images/yieldy-asynchronous-tasks.png)
+
+#### Yieldy Asynchronous Tasks and Threads
+
+Yieldy async tasks enable some concurrency, and these tasks are analogous to
+threads&mdash;well, non-preemptable threads running on a single-core machine:
+
+  * The `postTask()` callback is the thread entry point
+  * Calling `scheduler.yield()` or `scheduler.wait()` pauses the execution of
+    the current thread/task (similar to `Thread.yield()` and `Thread.sleep()` in
+    Java)
+  * Threads, like `postTask()` tasks, often have a modifiable priority
+  * Java's `Thread` also has the ability to get the current thread, which is
+    similar to the [`scheduler.currentTaskSignal`](../explainers/post-task-propagation.md)
+    proposal
+
+### The Challenge of Mixing Async APIs
+
+A major challenge with the [yieldy async task
+model](#yieldy-asynchronous-tasks) is that the browser doesn't currently know
+whether or not different browser tasks should be grouped together as the same
+userspace task. Aside from being conceptually undesirable, there is a practical
+implication around prioritization. Consider the following example:
+
+```javascript
+async function task() {
+  startWork();
+  await scheduler.yield();
+  doMoreWork();
+  let response = await fetch(myUrl);
+  let data = await response.json();
+  process(data);
+}
+
+scheduler.postTask(task, {priority: 'background'});
+```
+
+This would like something like this:
+
+![Mixed Yieldy Asynchronous Task Model](../images/mixed-yieldy-async-task.png)
+
+Even though the `fetch()` is conceptually part of the same task, we lose the
+task context (i.e. priority) since other async APIs are not yet
+"scheduler-aware". This means that only *part* of a yieldy task can be
+prioritized, which is something we want to address.
+
+From the browser's perspective, this results in the userspace task being spread
+across multiple [task
+sources](https://html.spec.whatwg.org/multipage/webappapis.html#task-source)[<sup>5</sup>](#notes).
+
+## Expanding the Task Model
+
+We would ideally like to create a unified userspace task model, which would
```

Comment:
This section did not go where I expected it to. The way this document was reading until this point, I was expecting the flow to be:

- Here's the general problem space
- Here's a simplified version
- Let's solve the simplified version!
- It turns out that solution can be generalized and works on the whole problem space!

Instead it seems like we're going with something like

- Here's the general problem space
- Here's a simplified version
- OK but we need to abandon the simplified version
- Now we're going to solve the general problem space

which raises the question as to why we bothered investigating the simplified version in the first place.

A flow that might make more sense is to not split into simplified model + expanded model, but instead just have a general section, something like "Problems to solve", which lists examples:

- Synchronous tasks
- Yieldly asynchronous tasks
- Mixing async APIs / entry point or continuation
- Multiple entry points
- Multiple tasks per browser task

where each example can just be about a specific scenario (which by its nature may or may not include the full complexity).

Then you could tackle "let's solve all these problems" in your forthcoming "creating a unified task model" section.

---

On the other hand, I'm not sure it's worth rewriting too much based on this feedback, as what you have works well and is relatively easy to follow. It was just surprising.

Reply:
Thanks for the feedback here, it's great to hear another perspective. I went back and forth on structure a few times. The challenge here I think, and what I was trying to avoid by introducing the simplified models, is not overwhelming readers with too much complexity too early on.

> This section did not go where I expected it to. The way this document was reading until this point, I was expecting the flow to be:
> 
> * Here's the general problem space
> * Here's a simplified version
> * Let's solve the simplified version!
> * It turns out that solution can be generalized and works on the whole problem space!
> 
> Instead it seems like we're going with something like
> 
> * Here's the general problem space
> * Here's a simplified version
> * OK but we need to abandon the simplified version

I didn't really see this approach as abandoning the simplified models. The simplified models are still valid subcases of any larger model, and IMO they serve the purpose of introducing two broad ways of thinking about tasks without too much complexity, however incomplete. Maybe it would be better to reframe these less formally (proposal below).

> * Now we're going to solve the general problem space
> 
> which raises the question as to why we bothered investigating the simplified version in the first place.
> 
> A flow that might make more sense is to not split into simplified model + expanded model, but instead just have a general section, something like "Problems to solve", which lists examples:
> 
> * Synchronous tasks
> * Yieldly asynchronous tasks
> * Mixing async APIs / entry point or continuation
> * Multiple entry points
> * Multiple tasks per browser task

I think my hangup is that to accurately model synchronous tasks, you need to include (4) and (5), and to model yieldy async tasks, you need to include (3), (4), and (5).

Maybe there's a middle ground where we don't call these simplified models, but still separate them from the challenges?

 * Overview
 * 2 general types of userspace tasks
    * Synchronous tasks
    * Yieldy async tasks
 * Challenges / Problems to solve / things to consider
    * Mixing async APIs / entry point or continuation
    * Multiple entry points
    * Multiple tasks per browser task
 * [Future work] Building a unified userspace task model

WDYT of this?

> where each example can just be about a specific scenario (which by its nature may or may not include the full complexity).
> 
> Then you could tackle "let's solve all these problems" in your forthcoming "creating a unified task model" section.
> 
> On the other hand, I'm not sure it's worth rewriting too much based on this feedback, as what you have works well and is relatively easy to follow. It was just surprising.

Reply:
> WDYT of this?

I like it!

Reply:
Okay, finally uploaded a new version. PTAL, thanks!